core/src/main/java/org/elasticsearch/action/support/replication/TransportReplicationAction.java

```diff
@@ -566,13 +528,70 @@ void finishOnRemoteSuccess(Response response) {
             }
         }
 
+        void retryBecauseUnavailable(ShardId shardId, String message) {
+            retry(new UnavailableShardsException(shardId, message + " Timeout: [" + internalRequest.request().timeout() + "], request: " + internalRequest.request().toString()));
+        }
+    }
+
+    /**
+     * Responsible for performing all operations up to the point we start sending requests to replica shards.
+     * When unavailable shards or any retry-able primary errors are encountered they are sent to {@link ReroutePhase}
+     * <p>
+     * Note that as soon as we start sending request to replicas, state responsibility is transferred to {@link ReplicationPhase}
+     */
+    final class PrimaryPhase extends AbstractRunnable {
+        private final InternalRequest internalRequest;
+        private final TransportChannel channel;
+        private final AtomicBoolean finished = new AtomicBoolean(false);
+        private volatile Releasable indexShardReference;
+        private final ClusterStateObserver observer;
+
+        PrimaryPhase(Request request, TransportChannel channel) {
+            this.observer = new ClusterStateObserver(clusterService, request.timeout(), logger);
+            final String concreteIndex;
+            if (resolveIndex()) {
```

Comment:
I wonder if we should resolve things again here. I think we should make it as simple as possible here. Maybe have a Primary request (or use internal request) which has the shardid already built into it. All we need to do in the action is lookup the index shard, see that we have it and do our thing.